File: training-data/go/valid_parentheses.go
package main

import "fmt"

func isValid(s string) bool {
	pairs := map[byte]byte{')': '(', ']': '[', '}': '{'}
	var stack []byte
	for i := 0; i < len(s); i++ {
		ch := s[i]
		if ch == '(' || ch == '[' || ch == '{' {
			stack = append(stack, ch)
		} else if open, ok := pairs[ch]; ok {
			if len(stack) == 0 || stack[len(stack)-1] != open {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}

func main() {
	fmt.Println(isValid("([{}])"))
	fmt.Println(isValid("([)]"))
}

File: training-data/r/valid_parentheses.R
is_valid <- function(s) {
  stack <- c()
  pairs <- c(")" = "(", "]" = "[", "}" = "{")
  chars <- strsplit(s, "")[[1]]
  for (c in chars) {
    if (c %in% c("(", "[", "{")) {
      stack <- c(stack, c)
    } else if (c %in% c(")", "]", "}")) {
      if (length(stack) == 0 || stack[length(stack)] != pairs[c]) {
        return(FALSE)
      }
      stack <- stack[-length(stack)]
    }
  }
  length(stack) == 0
}

print(is_valid("{[()]}"))
print(is_valid("{[(])}"))
